spatialos-sdk-tools/src/setup/main.rs

```diff
@@ -0,0 +1,106 @@
+    fs::create_dir_all(&bin_path)
+        .map_err(|_| format!("Failed to create {}", bin_path.display()))?;
+    fs::create_dir_all(&tmp_path)
+        .map_err(|_| format!("Failed to create {}", tmp_path.display()))?;
+
+    // Copy the contents of the schema-compiler/proto dir into the temp dir.
+    let proto_dir_glob = spatial_lib_dir.join("schema-compiler/proto/*");
+    for entry in glob::glob(proto_dir_glob.to_str().unwrap())?.filter_map(Result::ok) {
+        dir::copy(
+            &entry,
+            &tmp_path,
+            &CopyOptions {
+                overwrite: true,
+                ..CopyOptions::new()
+            },
+        )
+        .map_err(|_| {
+            format!(
+                "Failed to copy {} to {}",
+                entry.display(),
+                tmp_path.display()
+            )
+        })?;
+    }
+
+    // Run the schema compiler for each of the schema files in std-lib/improbable.
+    let schema_path_arg =
+        OsString::from("--schema_path=").tap(|arg| arg.push(&spatial_lib_dir.join("std-lib")));
+    let proto_out_arg = OsString::from("--proto_out=").tap(|arg| arg.push(&tmp_path));
+    let mut command = Command::new(&schema_compiler_path);
+    command
+        .arg(&schema_path_arg)
+        .arg(&proto_out_arg)
+        .arg("--load_all_schema_on_schema_path");
+
+    let schema_glob = spatial_lib_dir.join("std-lib/improbable/*.schema");
+    for entry in glob::glob(schema_glob.to_str().unwrap())?.filter_map(Result::ok) {
+        command.arg(&entry);
+    }
+
+    command
+        .status()
+        .map_err(|_| "Failed to compile schema files")?;
+
+    // Run protoc on all the generated proto files.
+    let proto_glob = tmp_path.join("**/*.proto");
+    let proto_path_arg = OsString::from("--proto_path=").tap(|arg| arg.push(&tmp_path));
+    let descriptor_out_arg = OsString::from("--descriptor_set_out=")
+        .tap(|arg| arg.push(&bin_path.join("schema.descriptor")));
+    let mut command = Command::new(&protoc_path);
+    command
+        .arg(&proto_path_arg)
+        .arg(&descriptor_out_arg)
+        .arg("--include_imports");
+    for entry in glob::glob(proto_glob.to_str().unwrap())?.filter_map(Result::ok) {
+        command.arg(&entry);
+    }
+
+    command.status().map_err(|_| "Failed to run protoc")?;
+
+    // Remove the temp directory once the setup process has finished.
+    fs::remove_dir_all(&tmp_path).map_err(|_| "Failed to remove temp dir")?;
+
+    Ok(())
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "setup", about = "Perform setup for a Rust SpatialOS project.")]
+struct Opt {
+    /// The path to your local installation of the SpatialOS SDK. If not specified,
+    /// the SPATIAL_OS_DIR environment variable instead.
+    #[structopt(long = "spatial-lib-dir", short = "l", parse(from_os_str))]
+    spatial_lib_dir: Option<PathBuf>,
+
+    /// The path the schema directory for the project.
+    #[structopt(parse(from_os_str))]
+    schema_dir: PathBuf,
```

Comment:
Would prefer if this was named to `output_dir` rather than `schema_dir`

Reply:
Done! Though in that case, does the current logic of appending `/bin` to `output_dir` make sense? I thought that was reasonable if the idea was that the user specifies their schema dir, we compile all their schemas, then output the compiled descriptor set to the `bin` dir inside their schema dir. If we're not explicitly asking for the schema dir, should I change it so that they specify the bin dir directly?

Reply:
Well you have have multiple input schema directories - which should compile down to one schema descriptor (not implemented yet, but a trivial thing to change). 

Ultimately you have some something like: 

```
cargo run --bin setup -- --schema-path=./my-user-path --schema-path=./my-other-schema-path --output-dir=./spatialos/schema
```

and the setup infers the std lib location from `SPATIAL_LIB_DIR` and globs any `schema-path` arguments

Reply:
Support for multiple `--schema-path` flags has been added, btw.